aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -128,6 +127,42 @@ impl<'e, E: ExecutorView> StorageAdapter<'e, E> {
             Ok((buf, buf_size))
         }
     }
+
+    fn get_any_resource_size_with_layout(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+        metadata: &[Metadata],
+        maybe_layout: Option<&MoveTypeLayout>,
+    ) -> PartialVMResult<Option<usize>> {
+        let resource_group = get_resource_group_member_from_metadata(struct_tag, metadata);
+        if let Some(resource_group) = resource_group {
+            let key = StateKey::resource_group(address, &resource_group);
+            let buf =
+                self.resource_group_view
+                    .get_resource_from_group(&key, struct_tag, maybe_layout)?;
+
+            let first_access = self.accessed_groups.borrow_mut().insert(key.clone());
+            let group_size = if first_access {
+                self.resource_group_view.resource_group_size(&key)?.get()
+            } else {
+                0
+            };
+
+            let buf_size = resource_size(&buf);
+            Ok(buf.map(|_| buf_size + group_size as usize))
+        } else {
+            let state_key = resource_state_key(address, struct_tag)?;
+            let buf_size = self
+                .executor_view
+                .get_resource_state_value_size(&state_key)?;
+            Ok(if buf_size == 0 {
```

Comment:
Hm, it seems it is safer to change `get_resource_state_value_size` to `Result<Option<u64>>` because probably in executor that is what we do internally
```rust
// in view.rs
if let ReadResult::ResourceSize(v) = res {
    v.unwrap_or(0)
} else {
    unreachable!("Read result must be ResourceSize kind")
}
```

Resources should not be zero sized but if this changes this will can become a bug